File: code/264.丑数-ii.cpp
/*
 * @lc app=leetcode.cn id=264 lang=cpp
 *
 * [264] 丑数 II
 */
#include <algorithm>
#include <vector>
using namespace std;
// @lc code=start
class Solution {
public:
    int nthUglyNumber(int n) {
        vector<int> rs(n, 1);
        int i2 = 0;
        int i3 = 0;
        int i5 = 0;
        for (size_t i = 1; i < n; i++) {
            rs[i] = min(min(rs[i2] * 2, rs[i3] * 3), rs[i5] * 5);
            if (rs[i] == rs[i2] * 2) {
                i2++;
            } 
            if (rs[i] == rs[i3] * 3) {
                i3++;
            }
            if (rs[i] == rs[i5] * 5) {
                i5++;
            }
        }
        return rs[n - 1];
    }
};
// @lc code=end

File: code/264.丑数-ii.py
#
# @lc app=leetcode.cn id=264 lang=python3
#
# [264] 丑数 II
#
import heapq
# @lc code=start
class Solution:
    def nthUglyNumber(self, n: int) -> int:

        nums = [1]
        eliminate = set()
        for _ in range(n - 1):
            curr = heapq.heappop(nums)
            
            for i in [2, 3, 5]:
                n = curr * i
                if n not in eliminate:
                    heapq.heappush(nums, n)
                    eliminate.add(n)

        return heapq.heappop(nums)
# @lc code=end

